internal/integration/k8s/usernamespace.go

```diff
@@ -0,0 +1,109 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+//go:build integration_k8s
+
+package k8s
+
+import (
+	"bytes"
+	"context"
+	_ "embed"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/siderolabs/talos/internal/integration/base"
+	"github.com/siderolabs/talos/pkg/machinery/client"
+	"github.com/siderolabs/talos/pkg/machinery/config/machine"
+)
+
+// UsernamespaceSuite verifies that a pod with user namespace works.
+type UsernamespaceSuite struct {
+	base.K8sSuite
+}
+
+//go:embed testdata/usernamespace.yaml
+var usernamespacePodSpec []byte
+
+// SuiteName returns the name of the suite.
+func (suite *UsernamespaceSuite) SuiteName() string {
+	return "k8s.UsernamespaceSuite"
+}
+
+// TestUsernamespace verifies that a pod with user namespace works.
+func (suite *UsernamespaceSuite) TestUsernamespace() {
+	if suite.Cluster == nil {
+		suite.T().Skip("without full cluster state reaching out to the node IP is not reliable")
+	}
+
+	if suite.Cluster.Provisioner() != base.ProvisionerQEMU {
+		suite.T().Skip("skipping usernamespace test since provisioner is not qemu")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	suite.T().Cleanup(cancel)
+
+	node := suite.RandomDiscoveredNodeInternalIP(machine.TypeWorker)
+
+	nodeCtx := client.WithNodes(ctx, node)
+
+	reader, err := suite.Client.Read(nodeCtx, "/proc/sys/user/max_user_namespaces")
+	suite.Require().NoError(err)
+
+	var maxUserNamespaces bytes.Buffer
+
+	_, err = maxUserNamespaces.ReadFrom(reader)
+	suite.Require().NoError(err)
+
+	suite.T().Logf("max_user_namespaces: ##%s##", maxUserNamespaces.String())
+
+	if strings.TrimSpace(maxUserNamespaces.String()) == "0" {
+		suite.T().Skip("skipping test since user namespace is disabled")
+	}
+
+	usernamespacePodManifest := suite.ParseManifests(usernamespacePodSpec)
+
+	suite.T().Cleanup(func() {
+		cleanUpCtx, cleanupCancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cleanupCancel()
+
+		suite.DeleteManifests(cleanUpCtx, usernamespacePodManifest)
+	})
+
+	suite.ApplyManifests(ctx, usernamespacePodManifest)
+
+	suite.Require().NoError(suite.WaitForPodToBeRunning(ctx, time.Minute, "default", "userns"))
+
+	processResp, err := suite.Client.Processes(nodeCtx)
+	suite.Require().NoError(err)
+
+	var sleepProcessPID int
+
+	for _, processInfo := range processResp.Messages {
+		for _, process := range processInfo.Processes {
+			if strings.Contains(process.Command, "sleep") {
+				sleepProcessPID = int(process.Pid)
+
+				break
+			}
+		}
+	}
+
+	suite.Require().NotZero(sleepProcessPID, "sleep process not found for usernamespace test")
+
+	reader, err = suite.Client.Read(nodeCtx, fmt.Sprintf("/proc/%d/status", sleepProcessPID))
+	suite.Require().NoError(err)
+
+	var processStatus bytes.Buffer
+
+	_, err = processStatus.ReadFrom(reader)
+	suite.Require().NoError(err)
+
+	suite.Require().NotContains(processStatus.String(), "Uid:	0	0	0	0", "sleep process should not have root UID in host namespace")
```

Comment:
this might be a bit weak check, I think we should better find a line with `Uid:`, split it up and look that first field is not 0 ?

Otherwise if the kernel format slightly changes, the check would pass, while it might be still uid 0